lib/sync/DatasetClient.js

```diff
@@ -0,0 +1,107 @@
+var syncUtil = require('./util');
+var datasets = require('./datasets');
+
+function generateDatasetClientId(datasetClient) {
+  return [datasetClient.datasetId, syncUtil.generateHash(datasetClient.queryParams), syncUtil.generateHash(datasetClient.metaData)].join("_");
+}
+
+/**
+ * Represent a dataset client that is managed by sync framework.
+ * @param {String} datasetId the id of the dataset
+ * @param {Object} opts options of the dataset client
+ * @param {Object} opts.queryParams the query parameters of the dataset client. Will be passed to the data handler to list data.
+ * @param {Object} opts.metaData the meta data of the dataset client. Will be passed to the data handler to list data.
+ * @param {Object} opts.props some properties of the dataset client. Mainly for record the internal status of the dataset client
+ * @param {Object} opts.config the configuration option of the dataset client
+ */
+function DatasetClient(datasetId, opts){
+  opts = opts || {};
+  this.datasetId = datasetId;
+  this.queryParams = opts.queryParams || {};
+  this.metaData = opts.metaData || {};
+  this.props = opts.props || {};
+  this.id = generateDatasetClientId(this);
+  this.config = opts.config || datasets.getDatasetConfig(datasetId);
+}
+
+/**
+ * Get the JSON presentation of the dataset client instance.
+ */
+DatasetClient.prototype.toJSON = function() {
+  return {
+    id: this.id,
+    datasetId: this.datasetId,
+    queryParams: this.queryParams,
+    metaData: this.metaData,
+    config: this.config,
+    props: this.props,
+    globalHash: this.globalHash
+  };
+};
+
+/**
+ * Check if the dataset client should be synced.
+ */
+DatasetClient.prototype.shouldSync = function() {
+  var self = this;
+  if (!self.props.syncLoopStart) {
+    // Dataset has never been synced before - do initial sync
+    return true;
+  }
+  // Time between sync loops has passed - do another sync
+  var lastSyncCmp = self.props.syncLoopEnd;
+  if (!lastSyncCmp) {
+    //the sync run hasn't been finished before, can't decide
+    return false;
+  }
+  var syncFrequency = self.config.syncFrequency * 1000;
+  var timeSinceLastSync = new Date().getTime() - lastSyncCmp;
```

Comment:
This could be `Date.now()` I think. I don't know if it's faster or anything but it might make it look a bit simpler. There's a few other places this could be replaced also.

```javascript
var timeSinceLastSync = Date.now() - lastSyncCmp;
```